compiler/rustc_data_structures/src/sync.rs

```diff
@@ -381,55 +385,106 @@ impl<K: Eq + Hash, V: Eq, S: BuildHasher> HashMapExt<K, V> for HashMap<K, V, S>
     }
 }
 
-#[derive(Debug)]
-pub struct Lock<T>(InnerLock<T>);
+pub struct Lock<T> {
+    single_thread: bool,
+    data: UnsafeCell<T>,
+    borrow: Cell<bool>,
+    mutex: RawMutex,
```

Comment:
Couldn't these two fields be merged into a single `AtomicBool`? The single-threaded version would become `borrow.as_ptr().replace(true)`, while the multi-threaded version would use `parking_lot_core::park` to handle waiting.

Reply:
Thanks! It makes sense, I'll try it soon